### PR TITLE
Fix to scroll when the height of the keyboard has changed

### LINKED ISSUE #338: Does not scroll to bottom when the height of the keyboard has changed
## General Information

* MessageKit Version:
0.10.1
* iOS Version(s):
iOS 11.1
* Swift Version:
Swift 4.0
* Devices/Simulators:
iPhone 8 (Simulators)
* Reproducible in ChatExample? (Yes/No):
yes

##  What happened?
messageCollectionView doesn't scroll to bottom when the height of the keyboard has changed.

message-kit

## What did you expect to happen?
I think that it is better to scroll `messageCollectionView` to bottom when the height of the keyboard has changed like messenger, iMessage and Slack.

### PR DESCRIPTION
**Pull request by hellokenta1024**
## What does this implement/fix? Explain your changes.

Hi. I want to try to resolve #338.
It's a simple improvement.
I just called `messagesCollectionView.scrollToBottom(animated: true)` in `handleKeyboardDidChangeState(_ notification: Notification)` when software keyboard has been found.

Please let me know if you need a flag like `scrollsToBottomOnKeyboardDidChangeState` etc.
Thanks

after-masssage-kit

## Does this close any currently open issues?

resolve #338 

## Where has this been tested?

**Devices/Simulators:** …
iPhone 8 (Simulator)

**iOS Version:** …
iOS 11.1

**Swift Version:** …
Swift 4.0

**MessageKit Version:** …
1.10.1

### CODE REVIEW

**Comment by SD10** [comment]
Hey @KEN-chan thanks for taking the time to submit this PR 👍 However, I'm not sure if this is the right approach for this even with a `scrollsToBottomOnKeyboardDidChangeState` flag.

Users may not want to scroll to the bottom on events like dismissing the keyboard or when the keyboard height changes. For example, a user could want to look at a previous message for reference when typing the current message.

I think this might work better by observing the `textViewDidChange` notification. 
What do you think? cc @cwalo @nathantannar4 @zhongwuzw

**Comment by nathantannar4** [comment]
A user could also use the `MessageInputBarDelegte` which can notify when the content size changes. A user could then call scrollToBottom. So I don’t think we need another flag

**Comment by hellokenta1024** [comment]
@SD10 @nathantannar4 
Thanks for your comments.
Indeed, `scrollToBottom` is not good.
How about scrolling the screen as much as the height of the keyboard changes?

I've fixed code. Could you check it??

fixed-message-kit

**Comment by SD10** [comment]
@KEN-chan Do you think we can add a flag for this so that the user can opt in to this behavior?

**Comment by nathantannar4** [comment]
@KEN-chan I notice in the gif that if the content offset is not at the max y (i.e., the user has scrolled up a bit) the rest of the messages still moves. Could you maybe add a check such that it only changes the contentOffset if the current offset is at the very bottom? @SD10 do you agree?

**Comment by SD10** [comment]
@nathantannar4 Hmmm, I’m not sure but I kind of like that it maintains the current contentOffset in the collection view. Any specific reason why you think we should constrain it to the bottom? 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by hellokenta1024** [comment]
@SD10 
> Do you think we can add a flag for this so that the user can opt in to this behavior?

Sure. 
I've added `scrollsOnKeyboardFrameChanges`.

**Comment by nathantannar4** [comment]
@SD10 I just thought the adjustment when in mid scroll looked odd and that the purpose of scrolling to the bottom would be to keep the bottom most message in the frame

**Comment by hellokenta1024** [comment]
@SD10 @nathantannar4 
I think it is natural because some major message app (Slack, FB Messenger) have this same behavior.
In the case of iMessage, it always scrolls to the bottom when the height of the keyboard changes.
What do  you think??

**Comment by SD10** [comment]
@KEN-chan I really like this functionality but I'm blocking merging it at the moment until I can give it a more in-depth review

**Comment by hellokenta1024** [comment]
@SD10 
I'm sorry to hurry you, could you review this when you have time?? 🙇

**Comment by SD10** [comment]
@KEN-chan This will be merged (I may have to change the name of the flag) but functionality seems good. I'm just trying to fix #301, #302 first and it's giving me some trouble

File: Sources/Controllers/MessagesViewController.swift [hunk 1]
```diff
@@ -45,6 +45,12 @@ open class MessagesViewController: UIViewController {
     ///
     /// The default value of this property is `false`.
     open var scrollsToBottomOnKeybordBeginsEditing: Bool = false
+    
+    /// A Boolean value that determines whether the `MessagesCollectionView` scrolls
+    /// according to the change of the height of the keyboard.
+    ///
+    /// The default value of this property is `false`.
+    open var scrollsOnKeyboardFrameChanges: Bool = false
```
Comment (SD10) [inline Sources/Controllers/MessagesViewController.swift]: Let's change `scrollsOnKeyboardFrameChanges` to `maintainPositionOnKeyboardFrameChange`

File: Sources/Controllers/MessagesViewController.swift [hunk 2]
```diff
@@ -281,17 +292,19 @@ fileprivate extension MessagesViewController {
...
-            messagesCollectionView.contentInset.bottom = bottomInset
-            messagesCollectionView.scrollIndicatorInsets.bottom = bottomInset
+            let afterBottomInset = keyboardEndFrame.height > keyboardOffsetFrame.height ? (keyboardEndFrame.height - iPhoneXBottomInset) : keyboardOffsetFrame.height
+            let differenceOfBottomInset = afterBottomInset - messageCollectionViewBottomInset
+            let contentOffset = CGPoint(x: messagesCollectionView.contentOffset.x, y: messagesCollectionView.contentOffset.y + differenceOfBottomInset)
+            
+            messageCollectionViewBottomInset = afterBottomInset
+            
+            if scrollsOnKeyboardFrameChanges {
+                messagesCollectionView.setContentOffset(contentOffset, animated: true)
```
Comment (SD10) [inline Sources/Controllers/MessagesViewController.swift]: Let's set `animated` to `false` here

**Review by SD10 — CHANGES_REQUESTED** [review]
@KEN-chan Sorry this took so long to review. We can move forward with this PR. If you could take care of the 2 changes I requested. I'll take care of the CHANGELOG entry for you 👍

File: Sources/Controllers/MessagesViewController.swift [hunk 3]
```diff
@@ -341,8 +336,8 @@ fileprivate extension MessagesViewController {
             
             messageCollectionViewBottomInset = afterBottomInset
             
-            if scrollsOnKeyboardFrameChanges {
-                messagesCollectionView.setContentOffset(contentOffset, animated: true)
+            if maintainPositionOnKeyboardFrameChange {
```
Comment (zhongwuzw) [inline Sources/Controllers/MessagesViewController.swift]: @SD10 Change from `maintainPositionOnKeyboardFrameChange ` to `maintainPositionOnKeyboardFrameChanged` would be better?

Comment (SD10) [reply Sources/Controllers/MessagesViewController.swift]: Sure, I'll change it when I add the CHANGELOG entry after I merge 👍

**Comment by SD10** [comment]
Thank you for contributing to MessageKit! I've invited you to join the MessageKit GitHub organization - no pressure to accept! If you'd like more information on what that means, check out our [contributing guidelines](https://github.com/MessageKit/MessageKit/blob/master/CONTRIBUTING.md) and join the [MessageKit Slack channel](https://join.slack.com/t/messagekit/shared_invite/enQtMjI4NzIzNzMyMzU0LTEyNmYzNzBlNjNiYjAwYjY1ZjhkMmVjNDE4OGNlMDY1OTRmNzVmMzI4MTg3OWQzZTM2NGU4NTU3ZDE5NzkwNzQ). Feel free to reach out if you have any questions! :smiley:

**Pull request merged**